File: functions/src/data/models/user-public-model.ts
import {UserPublic} from "../../domain/entities/user-public";
import {firestore} from "firebase-admin";
import DocumentData = firestore.DocumentData;


export class UserPublicModel extends UserPublic {
    static readonly kUid = "uid";
    static readonly kFirstName = "firstName";
    static readonly kLastName = "lastName";
    static readonly kAgoraUid = "agoraUid";

    constructor(
        uid: string,
        firstName: string,
        lastName: string,
        agoraUid: number,
    ) {
        super(uid, firstName, lastName, agoraUid);
    }

    static fromData(data:DocumentData | undefined) : UserPublicModel | null {
        if (!data) {
            return null;
        }
        return new UserPublicModel(
            data[UserPublicModel.kUid],
            data[UserPublicModel.kFirstName],
            data[UserPublicModel.kLastName],
            data[UserPublicModel.kAgoraUid],
        )
    }
}

File: flutter_app/lib/core/data/models/user_public_model.dart
import '../../domain/entities/user_public.dart';


class UserPublicModel extends UserPublic {
  /// Field names:
  static const String kUid = "uid";
  static const String kFirstName = "firstName";
  static const String kLastName = "lastName";


  UserPublicModel({required String uid, required String firstName, required String lastName})
      : super(uid: uid, firstName: firstName, lastName: lastName,);

  static UserPublicModel fromMap(map) {
    return UserPublicModel(
      uid: map[kUid],
      firstName: map[kFirstName],
      lastName: map[kLastName],
    );
  }

  static List<UserPublicModel> fromList(List list) {
    return list.map((data) => UserPublicModel.fromMap(data)).toList();
  }

  Map<String, dynamic> toMap() => {
    kUid: uid,
    kFirstName: firstName,
    kLastName: lastName
  };

}